### PR TITLE
[Linear cache] Add new UpdateResources method to update multiple resources at once without doing a stow update and limit unneeded allocations when processing delta watches

### PR DESCRIPTION
**Pull request by valerian-roche**
In use cases where linear cache is used with eds and delta watches, the user can currently only update the cache in two ways:
* full stow set (SetResources): this requires the user to build the entire list to then pass it to the cache, which then hashes everything to figure out what changed. This is very resource intensive (in our case ~1s of CPU). After that it will only check once for delta watches and return all changed endpoints at once
* provide detailed updates one by one (through UpdateResource and DeleteResource): this does not require the user to keep a full view of the data, and does not consume much cpu in hashing (only hashing the resources actually updated), but it will check all delta watches for each call, and will forward endpoints changed a key at a time (more traffic with clients)

In our case the main issue switching to the second part is the cpu usage (and time spent) on updates when updating multiple resources (though with still num_updates << total_keys, e.g. 50 vs 5000) while multiple delta watches exist (~20)
We then end up spending more than 120ms of CPU (for ~50 updates/deletes) with profiles like (this is aggregated for updates between 0 and 50):
image

In order to improve this, two changes are in this PR:
* Do not pre-allocate the slice of filtered resources in most cases
  * When the delta watch is non-wildcard, this is irrelevant to pre-allocate for the total list of resources, and most checks will return no change at all
  * When the delta watch is wildcard, pre-allocate when there is no version passed, but not in other cases when most resources have not been updated
* Provide a method to update/delete multiple resources at once. This allows only one iteration across watches, improving resource usage and limiting traffic between the control-plane and the clients

After the fix, runtime of the update (only includes iterating through calls to UpdateResource/DeleteResource before and on call to UpdateResources after) is greatly improved and stable (blue line below)
image

### CODE REVIEW

File: pkg/cache/v3/linear.go [hunk 1]
```diff
@@ -218,6 +218,33 @@ func (cache *LinearCache) DeleteResource(name string) error {
 	return nil
 }
 
+// UpdateResources updates a list of resources in the collection.
+// Resources provided with value == nil are considered as deleted.
```
Comment (s-matyukevich) [inline pkg/cache/v3/linear.go]: Not sure this good interface for a public facing API, this behavior isn't evident by looking at the method signature.  Maybe we can rewrite this to `UpdateResources(resources map[string]types.Resource, toDelete []string)` ?

File: pkg/cache/v3/delta.go [hunk 2]
```diff
@@ -59,19 +73,13 @@ func createDeltaResponse(ctx context.Context, req *DeltaRequest, state stream.St
...
 
-	// Compute resources for removal regardless of the request type
-	for name, prevVersion := range state.GetResourceVersions() {
-		// The prevVersion != "" check is in place to make sure we are only sending an update to the client once right after it is removed.
-		// If the client decides to keep the subscription we skip the add for every subsequent response.
-		if _, ok := resources.resourceMap[name]; !ok && prevVersion != "" {
-			toRemove = append(toRemove, name)
-		}
-	}
-
```
Comment (s-matyukevich) [inline pkg/cache/v3/delta.go]: Why did you delete this part?

Comment (valerian-roche) [reply pkg/cache/v3/delta.go]: I merged this part with the iterations above. In the case of non-wildcard, this avoids an entire map iteration that is already done (line 77). For wildcard code is just moved (line 56)

File: pkg/cache/v3/linear.go (hunk 1 above)
Comment (valerian-roche) [reply pkg/cache/v3/linear.go]: I was wondering on that. I agree toDelete as a separate slice might be more clear for users

Comment (valerian-roche) [reply pkg/cache/v3/linear.go]: Updated
Also added a bit more testing on mixing updates/deletes when watches are non-wildcard

**Review by s-matyukevich — APPROVED** [review]
